Panic on udpsession

I got the panic two times this week. How can i handle this?

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x93e4db]

goroutine 129363104 [running]:
github.com/miekg/dns.(*SessionUDP).RemoteAddr(...)
	/build/go/src/github.com/miekg/dns/udp.go:34
github.com/miekg/dns.(*response).RemoteAddr(0xc0025ebdd0, 0xc0001d6a80, 0xc000952d72)
	/build/go/src/github.com/miekg/dns/server.go:773 +0x5b

--- a/server.go
+++ b/server.go
@@ -82,6 +82,7 @@ type ConnectionStater interface {
 
 type response struct {
 	msg            []byte
+	closed         bool // connection has been closed
 	hijacked       bool // connection has been hijacked by handler
 	tsigTimersOnly bool
 	tsigStatus     error
@@ -749,6 +750,10 @@ func (w *response) WriteMsg(m *Msg) (err error) {
 
 // Write implements the ResponseWriter.Write method.
 func (w *response) Write(m []byte) (int, error) {
+	if w.closed {
+		panic("dns: Write called after Close")
+	}
+
 	switch {
 	case w.udp != nil:
 		n, err := WriteToSessionUDP(w.udp, m, w.udpSession)
@@ -768,31 +773,39 @@ func (w *response) Write(m []byte) (int, error) {
 		n, err := io.Copy(w.tcp, bytes.NewReader(m))
 		return int(n), err
 	default:
-		panic("dns: Write called after Close")
+		panic("dns: internal error: udp and tcp both nil")
 	}
 }
 
 // LocalAddr implements the ResponseWriter.LocalAddr method.
 func (w *response) LocalAddr() net.Addr {
+	if w.closed {
+		panic("dns: LocalAddr called after Close")
+	}
+
 	switch {
 	case w.udp != nil:
 		return w.udp.LocalAddr()
 	case w.tcp != nil:
 		return w.tcp.LocalAddr()
 	default:
-		panic("dns: LocalAddr called after Close")
+		panic("dns: internal error: udp and tcp both nil")
 	}
 }
 
 // RemoteAddr implements the ResponseWriter.RemoteAddr method.
 func (w *response) RemoteAddr() net.Addr {
+	if w.closed {
+		panic("dns: RemoteAddr called after Close")
+	}
+
 	switch {
 	case w.udpSession != nil:
 		return w.udpSession.RemoteAddr()
 	case w.tcp != nil:
 		return w.tcp.RemoteAddr()
 	default:
-		panic("dns: RemoteAddr called after Close")
+		panic("dns: internal error: udpSession and tcp both nil")
 	}
 }
 
@@ -807,13 +820,20 @@ func (w *response) Hijack() { w.hijacked = true }
 
 // Close implements the ResponseWriter.Close method
 func (w *response) Close() error {
-	// Can't close the udp conn, as that is actually the listener.
-	if w.tcp != nil {
-		e := w.tcp.Close()
-		w.tcp = nil
-		return e
+	if w.closed {
+		return nil
 	}
-	return nil
+	w.closed = true
+
+	switch {
+	case w.udp != nil:
+		// Can't close the udp conn, as that is actually the listener.
+		return nil
+	case w.tcp != nil:
+		return w.tcp.Close()
+	default:
+		panic("dns: internal error: udp and tcp both nil")
+	}
 }
 
 // ConnectionState() implements the ConnectionStater.ConnectionState() interface.

--- a/server.go
+++ b/server.go
@@ -779,10 +779,6 @@ func (w *response) Write(m []byte) (int, error) {
 
 // LocalAddr implements the ResponseWriter.LocalAddr method.
 func (w *response) LocalAddr() net.Addr {
-	if w.closed {
-		panic("dns: LocalAddr called after Close")
-	}
-
 	switch {
 	case w.udp != nil:
 		return w.udp.LocalAddr()
@@ -795,10 +791,6 @@ func (w *response) LocalAddr() net.Addr {
 
 // RemoteAddr implements the ResponseWriter.RemoteAddr method.
 func (w *response) RemoteAddr() net.Addr {
-	if w.closed {
-		panic("dns: RemoteAddr called after Close")
-	}
-
 	switch {
 	case w.udpSession != nil:
 		return w.udpSession.RemoteAddr()

--- a/server.go
+++ b/server.go
@@ -730,7 +730,7 @@ func (srv *Server) readUDP(conn *net.UDPConn, timeout time.Duration) ([]byte, *S
 // WriteMsg implements the ResponseWriter.WriteMsg method.
 func (w *response) WriteMsg(m *Msg) (err error) {
 	if w.closed {
-		panic("dns: WriteMsg called after Close")
+		return &Error{err: "WriteMsg called after Close"}
 	}
 
 	var data []byte
@@ -755,7 +755,7 @@ func (w *response) WriteMsg(m *Msg) (err error) {
 // Write implements the ResponseWriter.Write method.
 func (w *response) Write(m []byte) (int, error) {
 	if w.closed {
-		panic("dns: Write called after Close")
+		return 0, &Error{err: "Write called after Close"}
 	}
 
 	switch {

--- a/server.go
+++ b/server.go
@@ -729,6 +729,10 @@ func (srv *Server) readUDP(conn *net.UDPConn, timeout time.Duration) ([]byte, *S
 
 // WriteMsg implements the ResponseWriter.WriteMsg method.
 func (w *response) WriteMsg(m *Msg) (err error) {
+	if w.closed {
+		panic("dns: WriteMsg called after Close")
+	}
+
 	var data []byte
 	if w.tsigSecret != nil { // if no secrets, dont check for the tsig (which is a longer check)
 		if t := m.IsTsig(); t != nil {

--- a/server.go
+++ b/server.go
@@ -812,7 +812,13 @@ func (w *response) TsigStatus() error { return w.tsigStatus }
 func (w *response) TsigTimersOnly(b bool) { w.tsigTimersOnly = b }
 
 // Hijack implements the ResponseWriter.Hijack method.
-func (w *response) Hijack() { w.hijacked = true }
+func (w *response) Hijack() {
+	if w.closed {
+		panic("dns: Hijack called after Close")
+	}
+
+	w.hijacked = true
+}
 
 // Close implements the ResponseWriter.Close method
 func (w *response) Close() error {

--- a/server.go
+++ b/server.go
@@ -812,13 +812,7 @@ func (w *response) TsigStatus() error { return w.tsigStatus }
 func (w *response) TsigTimersOnly(b bool) { w.tsigTimersOnly = b }
 
 // Hijack implements the ResponseWriter.Hijack method.
-func (w *response) Hijack() {
-	if w.closed {
-		panic("dns: Hijack called after Close")
-	}
-
-	w.hijacked = true
-}
+func (w *response) Hijack() { w.hijacked = true }
 
 // Close implements the ResponseWriter.Close method
 func (w *response) Close() error {

--- a/server.go
+++ b/server.go
@@ -823,7 +823,7 @@ func (w *response) Hijack() {
 // Close implements the ResponseWriter.Close method
 func (w *response) Close() error {
 	if w.closed {
-		return nil
+		return &Error{err: "connection already closed"}
 	}
 	w.closed = true
 

--- a/server_test.go
+++ b/server_test.go
@@ -996,6 +996,15 @@ func TestResponseAfterClose(t *testing.T) {
 	})
 }
 
+func TestResponseDoubleClose(t *testing.T) {
+	rw := &response{
+		closed: true,
+	}
+	if err, expect := rw.Close(), "dns: connection already closed"; err == nil || err.Error() != expect {
+		t.Errorf("Close did not return expected: error %q, got: %v", expect, err)
+	}
+}
+
 type ExampleFrameLengthWriter struct {
 	Writer
 }

--- a/server_test.go
+++ b/server_test.go
@@ -725,7 +725,7 @@ func checkInProgressQueriesAtShutdownServer(t *testing.T, srv *Server, addr stri
 	srv.lock.RLock()
 	defer srv.lock.RUnlock()
 	if len(srv.conns) != 0 {
-		t.Errorf("TCP connection tracking map not empty after Shutdown; map still contains %d connections", len(srv.conns))
+		t.Errorf("TCP connection tracking map not empty after ShutdownContext; map still contains %d connections", len(srv.conns))
 	}
 }
 

--- a/server_test.go
+++ b/server_test.go
@@ -991,6 +991,9 @@ func TestResponseAfterClose(t *testing.T) {
 	testPanic("WriteMsg", func() {
 		rw.WriteMsg(new(Msg))
 	})
+	testPanic("Hijack", func() {
+		rw.Hijack()
+	})
 }
 
 type ExampleFrameLengthWriter struct {

--- a/server_test.go
+++ b/server_test.go
@@ -719,7 +719,11 @@ func checkInProgressQueriesAtShutdownServer(t *testing.T, srv *Server, addr stri
 	}
 
 	if eg.Wait() != nil {
-		t.Fatalf("conn.ReadMsg error: %v", eg.Wait())
+		t.Errorf("conn.ReadMsg error: %v", eg.Wait())
+	}
+
+	if len(srv.conns) != 0 {
+		t.Errorf("TCP connection tracking map not empty after Shutdown; map still contains %d connections", len(srv.conns))
 	}
 }
 

--- a/server_test.go
+++ b/server_test.go
@@ -988,12 +988,6 @@ func TestResponseAfterClose(t *testing.T) {
 	testPanic("Write", func() {
 		rw.Write(make([]byte, 2))
 	})
-	testPanic("LocalAddr", func() {
-		rw.LocalAddr()
-	})
-	testPanic("RemoteAddr", func() {
-		rw.RemoteAddr()
-	})
 }
 
 type ExampleFrameLengthWriter struct {

--- a/server_test.go
+++ b/server_test.go
@@ -988,6 +988,9 @@ func TestResponseAfterClose(t *testing.T) {
 	testPanic("Write", func() {
 		rw.Write(make([]byte, 2))
 	})
+	testPanic("WriteMsg", func() {
+		rw.WriteMsg(new(Msg))
+	})
 }
 
 type ExampleFrameLengthWriter struct {

--- a/server_test.go
+++ b/server_test.go
@@ -722,6 +722,8 @@ func checkInProgressQueriesAtShutdownServer(t *testing.T, srv *Server, addr stri
 		t.Errorf("conn.ReadMsg error: %v", eg.Wait())
 	}
 
+	srv.lock.RLock()
+	defer srv.lock.RUnlock()
 	if len(srv.conns) != 0 {
 		t.Errorf("TCP connection tracking map not empty after Shutdown; map still contains %d connections", len(srv.conns))
 	}

--- a/server_test.go
+++ b/server_test.go
@@ -993,9 +993,6 @@ func TestResponseAfterClose(t *testing.T) {
 	testPanic("WriteMsg", func() {
 		rw.WriteMsg(new(Msg))
 	})
-	testPanic("Hijack", func() {
-		rw.Hijack()
-	})
 }
 
 func TestResponseDoubleClose(t *testing.T) {

--- a/server_test.go
+++ b/server_test.go
@@ -983,9 +983,7 @@ func TestResponseAfterClose(t *testing.T) {
 	}
 
 	rw := &response{
-		tcp:        nil, // Close sets tcp to nil
-		udp:        nil,
-		udpSession: nil,
+		closed: true,
 	}
 	testPanic("Write", func() {
 		rw.Write(make([]byte, 2))

--- a/server_test.go
+++ b/server_test.go
@@ -972,27 +972,25 @@ func TestServerRoundtripTsig(t *testing.T) {
 }
 
 func TestResponseAfterClose(t *testing.T) {
-	testPanic := func(name string, fn func()) {
-		defer func() {
-			expect := fmt.Sprintf("dns: %s called after Close", name)
-			if err := recover(); err == nil {
-				t.Errorf("expected panic from %s after Close", name)
-			} else if err != expect {
-				t.Errorf("expected explicit panic from %s after Close, expected %q, got %q", name, expect, err)
-			}
-		}()
-		fn()
+	testError := func(name string, err error) {
+		t.Helper()
+
+		expect := fmt.Sprintf("dns: %s called after Close", name)
+		if err == nil {
+			t.Errorf("expected error from %s after Close", name)
+		} else if err.Error() != expect {
+			t.Errorf("expected explicit error from %s after Close, expected %q, got %q", name, expect, err)
+		}
 	}
 
 	rw := &response{
 		closed: true,
 	}
-	testPanic("Write", func() {
-		rw.Write(make([]byte, 2))
-	})
-	testPanic("WriteMsg", func() {
-		rw.WriteMsg(new(Msg))
-	})
+
+	_, err := rw.Write(make([]byte, 2))
+	testError("Write", err)
+
+	testError("WriteMsg", rw.WriteMsg(new(Msg)))
 }
 
 func TestResponseDoubleClose(t *testing.T) {

